Add a middleware to raise a special exception if the job is retried

thomasleese opened the pull request:
This makes it easier to filter out exceptions that occur but succeed in the next retry. The next part of this will be to configure `govuk_app_config` to ignore the new `RetryWorker` exception.

Currently we get a lot of errors from Sidekiq in Sentry when the job itself succeeds on the second or third so we don't really need to see these errors. This is especially prominent in [Publishing API](https://trello.com/c/Zslgi9sV/347-quieten-down-the-downstream-errors) and [Link Checker API](https://github.com/alphagov/link-checker-api/blob/master/app/workers/webhook_worker.rb#L36).

This does pose a problem where we may not see errors quick enough if a job takes 21 days to eventually fail, so an alternative might be to start showing errors after 2 or 3 retries.

kevindew commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+# Unrelease
```
missing a d

kevindew commented on lib/govuk_sidekiq/error_handling.rb:
```diff
@@ -0,0 +1,33 @@
+module GovukSidekiq
+  module ErrorHandling
+    class RetryWorker < StandardException
```
Should this be StandardError?

thomasleese replied:
yes it should! I'd accidentally left out `_spec` at the end of the tests file so they weren't even running...

kevindew commented on lib/govuk_sidekiq/error_handling.rb:
```diff
@@ -0,0 +1,33 @@
+        if last_try?(job)
+          raise e
+        else
+          raise RetryWorker.new(e)
```
I'm not actually sure you need to pass the exception through here I think Ruby might auto associate it as the cause. I find the Ruby exception and cause stuff a bit confusing though.

kevindew commented on lib/govuk_sidekiq/error_handling.rb:
```diff
@@ -0,0 +1,33 @@
+      def last_try?(job)
+        max_retries = job["retries"] || Sidekiq::Middleware::Server::RetryJobs::DEFAULT_MAX_RETRY_ATTEMPTS
+        retry_count = job["retry_count"] || 0
+        !job["retry"] || retry_count == max_retries - 1
```
are we at any risk of `job["retry"]` being nil by default and thus stopping retries?

thomasleese replied:
If it's `nil`, `job["retry"]` should be `true` meaning `last_try?` should be `true` and then the ordinary exception will be raised. I've actually refactored this logic to make it a bit clearer.

thomasleese commented:
Thanks for the review @kevindew, @tijmenb and @boffbowsh, I've refactored and changed how this works to only ignore certain kinds of exceptions, with an optional constructor to the middleware for apps to customise this list.

kevindew commented on lib/govuk_sidekiq/sidekiq_initializer.rb:
```diff
@@ -18,6 +18,7 @@ def self.setup_sidekiq(govuk_app_name, redis_config)
         config.server_middleware do |chain|
           chain.add Sidekiq::Statsd::ServerMiddleware, statsd: GovukStatsd, env: nil, prefix: "workers"
           chain.add GovukSidekiq::APIHeaders::ServerMiddleware
+          chain.add GovukSidekiq::ErrorHandling::RetryExceptionWrapperMiddleware
```
I think this class needs updating?

thomasleese replied:
Hmm yeah, wonder if there's a way I can test for that.

thomasleese commented:
@boffbowsh would you be able to take another look over this?

thomasleese commented:
Not actively working on this, will open again if it gets picked up later.
